### PR TITLE
Upgrade React

### LINKED ISSUE #941: Upgrade peerDep range to include React 17
Let's make the library usable with React 17 by updating the peerDeps range. Ideally we can allow either ^16.8 and ^17.

Another project at GitHub using PRC is running into issues with the [event delegation in older versions](https://reactjs.org/blog/2020/10/20/react-v17.html#changes-to-event-delegation) and is reliant on PRC upgrading to version 17 to move forward with shipping bug fixes.

### PR DESCRIPTION
**Pull request by emplums**
This PR changes the acceptable React version to be used as an external dependency to React `^17.0.0`.

I've also moved `react` from `dependencies` to `devDependencies`. Since it's marked as a `peerDependency` it shouldn't be in regular deps. I don't think we realized this because `rollup` is marking `react` as external (preventing it to be bundled with the application).

This change would require all applications using Primer React Components to upgrade to version 17 and should be released as a breaking change.

## Open Questions
Should we allow versions above 16.8 still? I'm pretty sure because I added `capture: true` to the two document listeners in the library, we should see the same behavior in environments using `^16.8.0` and environments using `^17.0.0`. It still feels a little weird to allow either version, but because `17.0.0` is relatively new, I wouldn't want someone to be prevented from using PRC because another dependency in their application requires something below 17 🤔 

Closes #941 

### Screenshots
Please provide before/after screenshots for any visual changes

### Merge checklist
- [ ] Added or updated TypeScript definitions (`index.d.ts`) if necessary
- [ ] Added/updated tests
- [ ] Added/updated documentation
- [ ] Tested in Chrome
- [ ] Tested in Firefox
- [ ] Tested in Safari
- [ ] Tested in Edge

Take a look at the [What we look for in reviews](https://github.com/primer/components/blob/main/CONTRIBUTING.md#what-we-look-for-in-reviews) section of the contributing guidelines for more information on how we review PRs.

### CODE REVIEW

File: src/SelectMenu/SelectMenu.js [hunk 1]
```diff
@@ -67,7 +67,7 @@ const SelectMenu = React.forwardRef(({children, initialTab, as, ...rest}, forwar
   // handles the overlay behavior - closing the menu when clicking outside of it
   useEffect(() => {
     if (open) {
-      document.addEventListener('click', onClickOutside)
+      document.addEventListener('click', onClickOutside, {capture: true})
```
Comment (emplums) [inline src/SelectMenu/SelectMenu.js:70]: See: https://reactjs.org/blog/2020/08/10/react-v17-rc.html#fixing-potential-issues

Comment (emplums) [reply src/SelectMenu/SelectMenu.js:70]: In the case of the SelectMenu and Details element we want to know about _any_ click in the document so we need to add `capture: true`

**Comment by emplums** [comment]
Note: I needed to remove the `files` key from our `now.json` - this started breaking today but is unrelated to the changes in this PR (I opened a separate PR to double check). Removing that key stops the Vercel deploys from failing and seems to make no changes to what is actually deployed 👍

**Review by colebemis — APPROVED** [review]
🚀

**Pull request merged**